src/lib/pbo/quadratization.jl

```diff
@@ -1,205 +1,178 @@
-# -*- :: Quadratization :: -*-
-
+# -*- :: Quadratization :: -*- #
 abstract type QuadratizationMethod end
 
-_aux(::Type{<:QuadratizationMethod}, ::Int) = 0
-_nst(::Type{<:QuadratizationMethod}, ::Int) = 0
-
-struct Quadratization{T<:QuadratizationMethod}
-    deg::Int # Initial Degree
-    aux::Int # Auxiliary variables
-    nst::Int # Non-Submodular Terms
+struct Quadratization{Q<:QuadratizationMethod}
+    stable::Bool
 
-    function Quadratization{T}(deg::Int) where {T<:QuadratizationMethod}
-        return new{T}(
-            deg,
-            _aux(T, deg),
-            _nst(T, deg),
-        )
+    function Quadratization{Q}(stable::Bool = false) where {Q<:QuadratizationMethod}
+        return new{Q}(stable)
     end
 end
 
 @doc raw"""
-    @quadratization(name, aux, nst)
-
-Defines a new quadratization technique.
-"""
-macro quadratization(name, aux, nst)
-    aux_func = if aux isa Integer
-        quote _aux(::Type{$(esc(name))}, ::Int) = $(aux) end
-    else
-        quote _aux(::Type{$(esc(name))}, deg::Int) = $(aux)(deg) end
-    end
-
-    nst_func = if nst isa Integer
-        quote _nst(::Type{$(esc(name))}, ::Int) = $(nst) end
-    else
-        quote _nst(::Type{$(esc(name))}, deg::Int) = $(nst)(deg) end
-    end
+    quadratize!(aux::Function, f::PBF{S, T}, ::Quadratization{Q}) where {S,T,Q}
 
-    return quote
-        struct $(esc(name)) <: QuadratizationMethod end
-
-        # Ancillary Variables
-        $(aux_func)
-
-        # Non-Submodular Terms
-        $(nst_func)
-    end
-end
-
-@doc raw"""
-    infer_quadratization(f::PBF)
-"""
-function infer_quadratization(f::PBF)
-    k = degree(f)
-
-    if k <= 2
-        return nothing
-    else
-        # Without any extra knowledge, it is better to
-        # quadratize term-by-term
-        return Quadratization{TBT}(k)
-    end
-end
-
-@doc raw"""
-    quadratize(aux::Function, f::PBF{S, T}, ::Quadratization) where {S, T}
-
-Quadratizes a given PBF, i.e. creates a function ``g \in \mathscr{F}^{2}`` from ``f \in \mathscr{F}^{k}, k \ge 3``.
+Quadratizes a given PBF in-place, i.e. applies a mapping ``Q : \mathscr{F}^{k} \to \mathscr{F}^{2}``.
 
 ```julia
 aux(::Nothing)::S
 aux(::Integer)::Vector{S}
 ```
-
-## Submodularity
-
-A function ``f : 2^{S} \to \mathbb{R}`` is said to be submodular if
-```math
-f(X \cup Y) + f(X \cap Y) \le f(X) + f(Y) \forall X, Y \subset S
-```
-""" function quadratize end
+""" function quadratize! end
 
 @doc raw"""
-    Quadratization{TBT}(::Int)
+    Quadratization{NTR_KZFD}(stable::Bool = false)
 
-Term-by-term quadratization. Employs other inner methods.
-"""
+NTR-KZFD (Kolmogorov & Zabih, 2004; Freedman & Drineas, 2005)
```

Comment:
The details on stable quadratization are mentioned in the quadratization book? If not they should go in the manuscript (you seem pretty big on it)

Reply:
It is documented here: https://github.com/psrenergy/ToQUBO.jl/blob/02c14765f3251b291fd04b27c989ebbec08d8cf6/src/attributes/compiler.jl#L74-L76

The idea is pretty simple: sorting the terms by length and then by lex order before proceeding to quadratization. Inside the algorithm, we sort the variables by lex too.

Reply:
In the quadratization section. Some people might want a stable version, and I haven't seen this anywhere else (in software or literature)

Reply:
Added to #63